pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,469 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	ctrlerror "github.com/gardener/gardener/pkg/controllerutils/reconciler"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// bastionEndpoints holds the endpoints the bastion host provides
+type bastionEndpoints struct {
+	// private is the private endpoint of the bastion. It is required when opening a port on the worker node to allow SSH access from the bastion
+	private *corev1.LoadBalancerIngress
+	//  public is the public endpoint where the enduser connects to establish the SSH connection.
+	public *corev1.LoadBalancerIngress
+}
+
+// Ready returns true if both public and private interfaces each have either
+// an IP or a hostname or both.
+func (be *bastionEndpoints) Ready() bool {
+	return be != nil && IngressReady(be.private) && IngressReady(be.public)
+}
+
+func (a *actuator) Reconcile(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	logger := a.logger.WithValues("bastion", client.ObjectKeyFromObject(bastion), "operation", "reconcile")
+
+	opt, err := DetermineOptions(bastion, cluster)
+	if err != nil {
+		return err
+	}
+
+	a.alicloudECSClient, err = getAliCloudECSClient(ctx, a.client, opt)
+	if err != nil {
+		return err
+	}
+
+	a.alicloudVPCClient, err = getAliCloudVPCClient(ctx, a.client, opt)
+	if err != nil {
+		return err
+	}
```

Comment:
As mentioned, use local variables to store the clients instead of the ones in the actuator.

Reply:
committed in b1de2d2b